### ISSUE TITLE
deno compile lacks support for jsx syntax

### ISSUE THREAD

**Issue by Ivan1931** [denoland/deno#9023]
I am trying to use the new `deno compile` on a project which makes use of `jsx` files. 

Currently I attempt to run `deno compile` on a `.tsx` file by running the following command:

```shell
deno compile --unstable server.tsx
```
which results in the following error message:

```shell
TS17004 [ERROR]: Cannot use JSX unless the '--jsx' flag is provided.
```

However it seems like the `--jsx` flag is not yet supported by `deno compile`. I see the following error message when attempting to add it:

```shell
deno compile --unstable --jsx server.tsx
error: Found argument '--jsx' which wasn't expected, or isn't valid in this context

USAGE:
    deno compile <source_file> --unstable

For more information try --help
```

It seems to me that there is no jsx support for deno compile - is this something which is planned in the future? Or hopefully I'm just not using the tool correctly and it's already supported :laughing: 

### Deno version

```shell
$ deno -V

deno 1.6.3
```

Please let me know if there is any other information required.

**Comment by kitsonk** [comment]
What happens when you do a `deno run server.tsx`?

**Comment by Ivan1931** [comment]
@kitsonk running `deno run server.tsx` works.

**Comment by coastalwhite** [comment]
I have done some investigation on this issue.

# TLDR
As far as I can tell, this is not a bug, but intended behaviour. Solution for this user would be adding a `tsconfig.json` file adding the `"jsx": "react"` option, this, however, does seem to be broken. A temporary solution (which I tested to work for me) could be using the `--no-check` flag as it seems to solve the problem. The would mean running `deno compile --unstable --no-check server.tsx`.

This has to do with (https://github.com/denoland/deno/discussions/7732) and raises some valid questions.

# Reproduction Steps

If you want to reproduce the issue yourself here are the steps:
1. We create a very small `.tsx` file. I have called it `test.tsx`. We fill it with some JSX code.
```typescript
// test.tsx

// Found at https://dev.to/craigmorten/writing-a-react-ssr-app-in-deno-2m7 on 13-01-2021
// @deno-types="https://raw.githubusercontent.com/Soremwar/deno_types/4a50660/react/v16.13.1/react.d.ts"
import React from "https://dev.jspm.io/react@16.13.1"

console.log(<i>Hi!</i>)
```

2. We run `deno --unstable compile path/to/test.tsx` or `cargo run -- --unstable compile path/to/text.tsx`, if you are in the repo. It will create the behaviour described.

# Cause

The cause of the issue described is as follows.
When we do the _check_ method of the _module graph_ (`/cli/module_graph.rs::Graph::check`) in we do __not__ automatically use the `"jsx": "react"` config options for _tsc_. When you append this option, the issue is gone.

The code then looks like this:
```rust
// /cli/module_graph.rs::Graph::check

/// Type check the module graph, corresponding to the options provided.
pub fn check(self, options: CheckOptions) -> Result<ResultInfo, AnyError> {
  let mut config = TsConfig::new(json!({
    // ... other config options
    "strict": true,
    "target": "esnext",
    "tsBuildInfoFile": "deno:///.tsbuildinfo",
    "jsx": "react",
  }));
  // ... rest of function.
}
```

Now the test file compiles and I assume the issue would also be gone for the __OP__.

This is, however, is not a good solution to the problem. It only indicates that the problem actually is this option not being present from the _tsc_ configuration. The bigger problem has to do with the code that follows that check:

```rust
// /cli/module_graph.rs::Graph::check

if options.emit {
  config.merge(&json!({
    // TODO(@kitsonk) consider enabling this by default
    //   see: https://github.com/denoland/deno/issues/7732
    "emitDecoratorMetadata": false,
    "jsx": "react",
    "inlineSourceMap": true,
    "outDir": "deno://",
    "removeComments": true,
  }));
} else {
  config.merge(&json!({
    "noEmit": true,
  }));
}
```

As you can see the the `jsx` option is sometimes added if the options emit is true. It is, however, __not__ set to true by callee of the _check_ method. As can be seen here:

```rust
// /cli/main.rs

let result_info = module_graph.clone().check(module_graph::CheckOptions {
  debug,
  emit: false,
  lib,
  maybe_config_path: program_state.flags.config_path.clone(),
  reload: program_state.flags.reload,
})?;
```

I am unsure why this is. I also don't think this is very important at the moment as we can also see that we should be able to fix it with the following line:

```rust
// /cli/module_graph.rs::Graph::check

let maybe_ignored_options =
      config.merge_tsconfig(options.maybe_config_path)?;
```

This would indicate that we can add a _tsconfig.json_ file stating `"jsx": "react"` (using the `--config [file]` flag). However, for some reason, this does not work and I will be submitting an issue for that. We can however still circumvent this with the `--no-check` flag and that seems to work.

# Discussion

Apart from the fact that this doesn't work, there should a discussion about this issue and its cause in my honest opinion. This is because as the __OP__ also describes later, he can `deno run` just fine. This to me seems kinda strange. Why is there extra configuration needed for `deno compile`. Why are the _tsconfig_ specifically used for the check so different? I can imagine that there are some differences, but why is _jsx_ mentioned for `deno run` but not for `deno compile`. This is also described in (https://github.com/denoland/deno/discussions/7732), and I suggest that we take another look at that discussion.

# Conclusion

This issue is probably not an actual bug, just (IMO weirdly) intended behaviour. For some reason, I can (at the moment) not fix this with the intended solution of using the ``--config`` flag. The `--no-check` does seem to work to solve the problem (for now).

**Comment by Ivan1931** [comment]
@coastalwhite Thanks for your investigation and the work around. 

Just to clarify that means running:

```shell
deno compile --unstable --no-check server.tsx
```

Running `--no-check` seems moderately scary especially if I hook this up to a CI process 😆  You're also saying that adding a `tsconfig.json` file with 

```json
...
"jsx": "react"
...
```

Should fix the original issue once #9014 is fixed?

**Comment by coastalwhite** [comment]
> Running `--no-check` seems moderately scary especially if I hook this up to a CI process 😆

I very much agree and that why I think issue #9104 is quite bad. Maybe for a temporary workaround, you can use:

```shell
deno --unstable lint server.tsx
deno --unstable compile --no-check server.tsx
```

This will pretty much simulate the `--no-check` behaviour that is very weird with the `deno compile` subcommand.

> You're also saying that adding a `tsconfig.json` file with
> 
> ```json
> ...
> "jsx": "react"
> ...
> ```
>
> Should fix the original issue once #9014 is fixed?

I think you mean #9104, but yes. This is written as such in the source code and is also documented as such, although personally to me this seems quite convoluted.

**Comment by coastalwhite** [comment]
Having looked further into #9104, it was just some syntax error by me. Here is the solution with the tsconfig.

Write a `tsconfig.json` with content:
```json
{
	"compilerOptions": {
		"jsx": "react"
	}
}
```

Then run (flag order is important here):
```shell
deno compile --unstable --config tsconfig.json server.tsx
```

This should fix your problem.
